src/py/flwr_experimental/baseline/dataset/pytorch_cifar_partitioned.py

```diff
@@ -0,0 +1,218 @@
+    augment_factor: int,
+    augment_transform: torchvision.transforms,
+    save_thumbnails_dir: str = "",
+) -> XY:
+    """Augments the dataset augment_transform
+
+    Dataset is augmented to len(original_dataset)*augment_factor
+    using augment_transform. The original dataset will be a subset
+    of the new one.
+
+    Parameters
+    ----------
+    original_dataset : torch.utils.data.Dataset
+        Original PyTorch Dataset to be augmented
+    augment_factor : int
+        Number of aumented copies generated from each original sample.
+        This will multiply the final length of the dataset by augment_factor.
+    augment_transform : torchvision.transforms
+        Composition of transforms used to augment the images.
+        Last transform in this composition does not need to be a ToTensor.
+    save_thumbnails_dir
+        Directory where to save thumbnails. One jpg image will be created for each original image
+        with a augment_factor thumbnails generated.
+
+    Returns
+    -------
+    XY
+        A tuple containing the new augmented dataset
+    """
+    if save_thumbnails_dir:
+        Path(save_thumbnails_dir).mkdir(parents=True, exist_ok=True)
+
+    combined_transform = torchvision.transforms.Compose(
+        [augment_transform, torchvision.transforms.ToTensor()]
+    )
+
+    augmented_imgs = np.empty(
+        (len(original_dataset), augment_factor, 3, 32, 32), dtype=np.float32
+    )
+    augmented_labels = np.empty((len(original_dataset), augment_factor), dtype=np.int32)
+
+    print("Generating augmented images...")
+    for idx, (img, label) in enumerate(tqdm(original_dataset)):
+        augmented_imgs[idx, 0, :, :, :] = torchvision.transforms.ToTensor()(img)
+        augmented_labels[idx, :] = label
+        for j in range(1, augment_factor):
+            augmented_imgs[idx, j, :, :, :] = combined_transform(img)
+        if isdir(save_thumbnails_dir):
+            tmp = [*augmented_imgs[idx, :, :, :, :]]
+            tmp2 = np.concatenate(tmp, axis=-1)
+            save_image(
+                from_numpy(tmp2), join(save_thumbnails_dir, f"cifar10_{idx}.jpg")
+            )
+
+    # Rearrange dataset. Every multiple of augment_factor is an original image.
+    X = augmented_imgs.reshape(-1, 3, 32, 32)
+    Y = augmented_labels.reshape(-1)
+    return (X, Y)
+
+
+def generate_partitioned_dataset_files(
+    dataset: XY,
+    len_partitions: int,
+    nb_partitions: int,
+    data_dir: str,
+) -> None:
+    """Generates a set of tensor files contaning partitions of dataset.
+
+    Parameters
+    ----------
+    dataset:
+        Dataset in the form XY to be partitioned.
+    len_partitions : int
+       Number of samples inside a partition.
+    nb_partitions : int
+        Total number of partitions to be generated.
+    data_dir: str
+        Path to directory where the partition dataset files will be stored.
+    """
+
+    X, Y = dataset
+
+    X, Y = shuffle(X, Y)
+    X, Y = sort_by_label(X, Y)
```

Comment:
I think this line is not required because of the next line the sorting order will be overridden anyway.

Reply:
I guess it depends on what we want to do and if I understood the `sort_by_label` correctly. I think `sort_by_label` will not shuffle before reordering, right? `dataset` is highly correlated, meaning that for each original image found in the dataset, the following ones will be augmented version of those. I'm not sure, but if `sort_by_label` works as I expect, then the first client would have all the original images. This would actually be periodic. Every N clients would have only original images. On the other hand this could be useful because we could launch only clients with those `id`s and see if augmentation has a bad effect.

Reply:
I might have been a bit unclear. The shuffle is totally fine and reasonable. What I mean is:

> line 136 => X, Y = sort_by_label(X, Y)
> line 137 => X, Y = sort_by_label_repeating(X, Y)

The `sort_by_label_repeating` will internally anyway call `sort_by_label` first thing.
https://github.com/adap/flower/blob/main/src/py/flwr_experimental/baseline/dataset/dataset.py#L46

Therefore the line 136 is not needed/redundant.

Reply:
Oh, I see. Okay. I'll change this

Reply:
I see. I missed the connection with 137. I'm removing 136

Reply:
Solved in https://github.com/adap/flower/commit/5235c38e8259f44f47f609a2a1ee36bbda7c1864